lib/secure_headers/headers/content_security_policy.rb

```diff
@@ -108,15 +109,32 @@ def build_media_type_list_directive(directive)
     def build_source_list_directive(directive)
       source_list = @config.directive_value(directive)
       if source_list != OPT_OUT && source_list && source_list.any?
-        minified_source_list = minify_source_list(directive, source_list).join(" ")
+        cleaned_source_list = clean_malformatted_sources(directive, source_list)
+        minified_source_list = minify_source_list(directive, cleaned_source_list).join(" ")
+        [symbol_to_hyphen_case(directive), minified_source_list].join(" ").strip
+      end
+    end
 
-        if minified_source_list =~ /(\n|;)/
-          Kernel.warn("#{directive} contains a #{$1} in #{minified_source_list.inspect} which will raise an error in future versions. It has been replaced with a blank space.")
+    # Private: Calculates a modified version version of source_list:
+    # - where any expression containing a deprecated character (semicolon or newline) has been split by that
+    # character, and
+    # - any empty string expression is removed.
+    def clean_malformatted_sources(directive, source_list)
+      cleaned_source_list = []
+      semicolon_warned_yet = false
+      source_list.map do |expression|
```

Comment:
This isn’t a map operation since it doesn’t have a meaningful array result. However we could do this using a [reduce](https://apidock.com/ruby/Enumerable/reduce) pattern, what do you think? `reduce` is a weird name for the operation in this context. Ruby also offers `inject` which is the same thing just aliased.

```ruby
cleaned_source_list = source_list.inject([]) do |arr, source|
  # … add elements to `arr` as needed.
end
```

If we like the code as-is, we could just change this to `each`.